### ISSUE TITLE
Bad response from captcha gateway. we got 404

### ISSUE THREAD

**Issue by r-martins** [ProxiBlue/reCaptcha#39]
I'm implementing Firecheckout and I'm getting Invalid Captcha when trying to move from billing address to the next step.
<img width="1496" alt="image" src="https://user-images.githubusercontent.com/191149/64586296-3dd44500-d3df-11e9-8e35-e0b424a1c77e.png">

When debugging, I found that it's doing a request to `https://www.google.com:443//recaptcha/api/verify` which returns 404.

I'm still investigating, but it seems this is no longer supported.

I'm using invisible captcha, and the event that is failing is `register_during_checkout`.

Does it make sense?

**Comment by ProxiBlue** [comment]
Ok, that url shoudl be:

https://www.google.com/recaptcha/api/siteverify

Checking to see why / where it is generated, and doing an update now

**Comment by ProxiBlue** [comment]
Ah, no wait, there seems to be two URLS, I looked potetially at the wrong one:

https://github.com/ProxiBlue/reCaptcha/blob/master/app/code/community/ProxiBlue/ReCaptcha/Helper/Data.php#L18

**Comment by ProxiBlue** [comment]
I think the issue is here: https://github.com/ProxiBlue/reCaptcha/blob/master/app/code/community/ProxiBlue/ReCaptcha/Model/Recaptcha.php#L133

In that code, it is a fallback, in case there is no param called 'g-recaptcha-response' response in the POST response, so it attempts to build the result.

I suspect this code has not been enacted by someone in some time, and it used another path to enact verification (why, I have no idea right now)

The does point to potentially a problem in the form posted data, and no recapctha field is present.

I will try and re-enact this, and check if that code works/fix it as the fallback.

**Comment by ProxiBlue** [comment]
You can try and change that line to be:

$path = ProxiBlue_ReCaptcha_Helper_Data::RECAPTCHA_SITEVERIFY_PATH;

and see if it works, but from the resulting code, it lookslike 'verify' url path used to come back with a different type of result.

Google must have dropped that path at some point (possible with v1)

**Comment by ProxiBlue** [comment]
i very much suspect this code is legacy, involving v1 recaptcha. I think you need to check what / where the recapctha response is that is supposed to be in the POST request, it seems to be missing, hence why you are ending up in that conditional.

I will have a look to what goes on there, as it look slike this code needs adjustment, and is very likely part of removed v1, that was missed to be removed.

**Comment by r-martins** [comment]
I tried to change to SITEVERIFY_PATH but got the same result in the frontend (Incorrect Captcha).

![magenteiro____www_magenteiro__-_____app_code_community_ProxiBlue_ReCaptcha_Model_Recaptcha_php](https://user-images.githubusercontent.com/191149/64656247-37db7400-d472-11e9-965e-8bed76584914.jpg)

On the log:

> 2019-09-10T22:56:03+00:00 DEBUG (7): Form ID: register_during_checkout=>result is : {
>   "success": false,
>   "error-codes": [
>     "missing-input-response",
>     "missing-input-secret"
>   ]
> }
>

**Comment by ProxiBlue** [comment]
Can you confirm there is a recaptcha response in the POST data?

$request->getParam('g-recaptcha-response')

**Comment by r-martins** [comment]
It was null/empty.
![Finalize_sua_inscrição_e_comece_a_assistir_agora](https://user-images.githubusercontent.com/191149/64657104-3a8b9880-d475-11e9-979f-96bb9530b7b4.jpg)

I tried to mannually execute `grecaptcha.execute()` on that page's console. The image challenge was displayed; I selected the image; got a `uncaught exception: null` on the console, but was able to move forward to the next step.

After doing that, the second request contains the g-recaptcha-response.
![Finalize_sua_inscrição_e_comece_a_assistir_agora](https://user-images.githubusercontent.com/191149/64657171-93f3c780-d475-11e9-9f77-cb7158550c6c.jpg)

It seems that the `app/design/frontend/base/default/template/captcha/recaptcha.phtml` event listeners are not listening to the firecheckout form.

**Comment by ProxiBlue** [comment]
Ok, progress. So issue might be how firecheckout has the form defined. 

I generally only use this with core magento checkout, so the js is a bit specific.

Stick a breakpoint here:

https://github.com/ProxiBlue/reCaptcha/blob/master/app/design/frontend/base/default/template/captcha/recaptcha.phtml#L18

Is the form element found? defined? It is likely you simply need to extend this .phtml, into theme, and change how that form for submitting is defined.

The form submit event is capture here:

https://github.com/ProxiBlue/reCaptcha/blob/master/app/design/frontend/base/default/template/captcha/recaptcha.phtml#L53

so stick a break point in there to see what the code is running, doing.

**Comment by ProxiBlue** [comment]
Depending on checkout options (like guest allowed etc), the submit could also be captured here, so again use a breakpoint and see if this is enacted.

https://github.com/ProxiBlue/reCaptcha/blob/master/app/design/frontend/base/default/template/captcha/recaptcha.phtml#L76

Possible issue is with the ``` if (this.validator && this.validator.validate()) {``` conditional?

**Comment by ProxiBlue** [comment]
> when trying to move from billing address to the next step.

just noticed that in your original question.

so could be here as well:

https://github.com/ProxiBlue/reCaptcha/blob/master/app/design/frontend/base/default/template/captcha/recaptcha.phtml#L89

With potential of having 3 recapctahs on the one screen (guest/register/address) there are multiple submit traps!

You might have to find an event in firecheckout to attach to, via some custom code, and run something similar to:

```
VarienForm.prototype.submit = VarienForm.prototype.submit.wrap(function (parentMethod) {
                    window.parentMehod = parentMethod;
                    if (this.validator && this.validator.validate()) {
                        // this captures the main login form on checkout, potentially others on site
                        // since there is the potential to other areas using this execute, it is not specific to
                        // a particular captcha
                        grecaptcha.execute();
                    }
                    return false;
                });
```

**Comment by r-martins** [comment]
Hello again @ProxiBlue.
Thank you very much for your help on this one.

I've rolled back the other RECAPTCHA_VERIFY_PATH, because I think it's not used anyway in my case.

I've copied the `/captcha/recaptcha.phtml` to my theme, and added the following right after `<?php $handle = $this->getLayout()->getUpdate()->getHandles(); ?>`:

```
<?php if(in_array('firecheckout_index_index', $handle)): ?>
document.observe('firecheckout:updateBefore', function(url, parameters) {
    grecaptcha.execute();
});
<?php endif;?>
```

This would execute `grecaptcha.execute();` before moving forward on every step of the checkout.

Thank you very much for the help. You are always helping me when I struggle with your module. :)
@ProxiBlue, how can I make a donation for you/the project?
